tests/macro-errors.rs

```diff
@@ -5,8 +5,15 @@
 // <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
 // option. This file may not be copied, modified, or distributed
 
+extern crate version_check;
+use version_check::Channel;
+
 #[test]
 fn ui() {
-    let t = trybuild::TestCases::new();
-    t.compile_fail("tests/ui/*.rs");
+    if Channel::read().unwrap().is_nightly() {
+        eprintln!("nightly build detected, skipping ui tests");
```

Comment:
When this nightly will go to beta, beta will also fail :-/

Reply:
And when beta go stable we're going to have a crash on stable \***bad sob**\*. Well, are we going to skip beta too?

Reply:
I don't know what to do, but I think at the end, the test will be done everywhere except on the smallest supported rust version.

Reply:
Maybe we can for the moment have nightly and beta as allow_failure, and when stable will break, we will update the test, remove the allow failure on stable and nightly, and remove the tests on the smallest supported rust version (i.e. this version will just compile)

Reply:
I don't like that, but I don't have a better proposition.

Reply:
Honestly, the more I think about the idea of skipping tests on certain channels the less I like it. I'm not sure what your policy on supporting nightly is, but I guess from the existence of `nightly` feature you claim to provide at least partial support, i.e. "it works, until these compiler guys break something again". Such a support implies that the errors that *must* happen will  actually happen.

If we chose to ignore these tests we effectively say "sorry, nightly guys, but we don't check that all the parts are *actually* work", and while this still would work it's not the best support we can provide. Worse yet, should something *actually* break in ui-tests - we won't even know about it.

I propose one of these choices instead
- Separate test suits for every channel, i.e `ui-stable/ui-beta/ui-nightly` subfolders. Only 4/30 errors are different, it's not too much to look after. The main drawback here - we'll have to support additional tests, at least until `stable/nightly/beta` are in sync. The main advantage - should something actually break and we know about it the very instant.
- We may use `allow_failures`, so travis won't bother you with this ui stuff and we still will be able to periodically check whether the error has changed. Drawback - we won't be notified, we'll have to *check* manually. (The question - is there a way to configure travis to notify once error changed? That would have annihilate the drawback).

What do you think?

Reply:
Can't we switch the .stderr file according to the rust version?

Reply:
Exactly my first proposal - pick tests depending on channel, details may differ: 
- separate folders
- separate folders + symlinks
- explicit code in `macro_errors.rs` that picks needed stuff

I'm about to make few commits so you could judge for yourself

Reply:
And [here is](https://github.com/CreepySkeleton/structopt/tree/update_deps2) another option

Reply:
* is it possible without duplication the input files?
* can't we switch on the rust version (>= 1.39, < 1.39) instead of the channel?
* If duplication is not possible, can we have a test suite of common, and another spited by version

Reply:
>     * is it possible without duplication the input files?

Symlinks?

Reply:
>     * can't we switch on the rust version (>= 1.39, < 1.39) instead of the channel?

Good idea, we definitely can

Reply:
>     * If duplication is not possible, can we have a test suite of common, and another spited by version

[Like this?](https://github.com/CreepySkeleton/structopt/tree/update_deps2)

Reply:
The `update_deps2` version seems perfect, with 2 sets switched by rust version.